### DISCUSSION TITLE
아이템 1. 생성자 대신 정적 팩터리 메서드를 고려하라

### DISCUSSION THREAD

**Discussion by Irisation23** [Study-2-Effective-Java/Effective-Java#17] [3. 과제] [answered]
# [아이템 1] - 생성자 대신 정적 팩터리 메서드를 고려하라.

> 🚨 시작에 앞서 **정적 팩터리 메서드**는 **팩터리 메서드**와 다르다.

# 팩터리 메서드 ⁉️
1. 객체 생성을 위한 인터페이스를 정의하지만 인터페이스를 구현하는 클래스가 인스턴스화 할 클래스를 결정하도록 함.
2. 팩토리 메서드를 사용하면 클래스가 인스턴스화를 하위 클래스로 위임할 수 있다. - GoF

# 정적 팩터리 메서드 ⁉️
- 객체 생성을 `캡슐화` 하는 방법중 하나임.
- 팩터리 메서드가 없으면 클래스의 생성자를 직접 호출하면 됨. `Foo foo = new Foo();`
- 정적 팩터리 메서드를 사용한다면 `Foo foo = Foo.getInstance()` 를 통해 호출 가능해짐.
- 정적 팩터리 메서드는 단순히 클래스의 인스턴스를 반환하는 정적 메서드임.
  - 그렇다면 그냥 단순히 클래스를 생성 하는것이 낫지않냐?
  - 왜 이걸 사용해야 하는가?

# 정적 팩터리 메서드의 장점 ✨
# 1. **이름을 가질 수 있다.**

```java
new BigInteger(int, int, Random);
BigInteger.probablePrime;
```
- 위의 코드는 `값이 소수인 BigInteger를 반환한다` 라는 뜻을 지닌다.
- 두 코드다 같은 행위를 한다.
- 즉 이름을 가질 수 있다는 의미는 **반환될 객체의 특성을 설명함에 유용해징다**라고 해석할 수 있다.

# 2. 호출될 때마다 인스턴스를 새로 생성하지 않아도 된다.
- 불필요한 객체 생성을 피할 수 있다.
- 반복되는 요청에 같은 객체를 반환하는 방식으로 동작한다.
- 언제 어느 인스턴스를 살아 있게 할지 통제가 가능해진다. `인스턴스 통제(instance-cotrolled)` 클래스라 한다.
> 🧐 통제의 이유는 무엇일까?
> - 인스턴스를 통제하면 클래스를 싱글턴으로 만들 수도, 인스턴스화 불가로 만들 수 있다.
> - 또한 불변 값 클래스에서 동치인 인스턴스가 단 하나뿐임을 보장할 수 있다.
    > `(a == b 일 때만 a.equals(b) 가 성립`
> - 이는 플라이웨이트 패턴의 근간이 되며, 열거 타입의 인스턴스가 단 하나만 됨을 보장할 수있다.

- [플라이웨이트 패턴 - 참조자료](https://lee1535.tistory.com/106)

# 3. 반환 타입의 하위 타입 객체를 반환할 수 있는 능력이 있다.
- 해당 기능은 객체지향 프로그래밍을 구성하는 `다형성`을 보여주는 예이다.
- 정적 팩터리 메서드의 리턴 타입을 인터페이스로 할 경우 하위 타입 객체를 반환 할 수 있다.
- 이는 주로 `동반 클래스`에서 볼 수 있다.

> 🧐 `동반 클래스`는 무엇일까?
>  - 동반 클래스: 자바 1.8 이전에는 인터페이스에 정적 메서드를 선언할 수 없었다.
>  - 따라서 인터페이스에 기능을 추가하기 위해서는 `동반 클래스` 라는 것을 만들어 그 안에 정적 메서드를 추가했다.
```java
public interface Animal { 
    class Animals { // 이 Animals 가 바로 동반 클래스!
        static Animal of(String legs) {
          if (Integer.parseInt(legs) == 4) {
            return new Dog();
          } else if (Integer.parseInt() > 4) {
            return new Octopus();
          }
        }
    } 
}
```
- 다음은 Collections 클래스의 정적 팩토리 메서드 unmodifiableList의 모습이다.
```java
public static <T> List<T> unmodifiableList(List<? extends T> list) {
    return (list instanceof RandomAccess ? 
            new UnmodifiableRandomAccessList<>(list) : 
            new UnmodifiableList<>(list));
}
```
- 이와 같이 팩터리 메서드를 통해 메서드의 반환 타입은 `List` 이지만 실제로는 `List`의 하위 객체를 반환 시킬 수 있다.
- 사용자들은 더 이상 해당 인터페이스의 구현체를 일일이 알아 볼 필요가 없다.
- 구현 클래스를 공개하지 않고도 그 객체를 반환할 수 있어서 API를 작게 유지할 수 있다.
- 해당 기능은 인터페이스 기반 프레임 워크를 만드는 핵심 기술이다.
- 대표적인 프레임워크로는 `Spring FrameWork`와 `JDBC`가 있다.
  - [Spring 프레임워크의 정적 팩터리 메서드 적용 예제 - by Baeldung](https://www.baeldung.com/spring-beans-factory-methods)

# 4. 입력 매개변수에 따라 매번 다른 클래스의 객체를 반환할 수 있다.
- 반환 타입의 하위 타입이라면 어떤 클래스의 객체를 반환해도 상관없다.
```java
public static <E extends Enum<E>> EnumSet<E> noneOf(Class<E> elementType) {
        Enum<?>[] universe = getUniverse(elementType);
        if (universe == null)
            throw new ClassCastException(elementType + " not an enum");

        if (universe.length <= 64)
            return new RegularEnumSet<>(elementType, universe);
        else
            return new JumboEnumSet<>(elementType, universe);
    }
```
- 메서드의 내부 구현이 변경되어도, 반환타입만 같다면 사용자에게 영향을 끼치지 않는다.
# 5. 정적 팩터리 메서드를 작성하는 시점에는 반환할 객체의 클래스가 존재하지 않아도 된다.
[의존 객체 주입의 내용과 오버랩 되어서 링크를 달았습니다.](https://github.com/orgs/Study-2-Effective-Java/discussions/18)

---
# 정적 팩터리 메서드의 단점 💥
# 1. 상속을 하려면 public이나 protected 생성자가 필요하니 정적 팩터리 메서드만 제공하면 하위 클래스를 만들 수 없다.
- 상속을 위해서는 하위 클래스에서 사용할 수 있는 생성자가 필요하다.
- private 생성자를 통해 외부 생성을 막고, 정적 팩터리 메서드만을 사용하여 인스턴스를 반환할 경우, 하위 클래스를 만들 수 없다.

# 2. 정적 팩터리 메서드는 프로그래머가 찾기 어렵다.
- 생성자는 이름이 클래스의 이름과 같아서 명확히 알 수 있다.
- 하지만 정적 팩터리 메서드는 다른 메서드와 섞여 찾기 어려울 수 있다.
- 생성자처럼 API 설명에 명확히 드러나지 않으니 사용자는 정적 팩터리 메서드 방식 클래스를 인스턴스화할 방법을 직접 알아내야 한다.

---

# 정적 팩터리 메서드의 명명 방식
- from: 매개변수 하나를 받아 해당 타입의 인스턴스를 반환하는 형변환 메서드
```java
// intance 를 Date로 변환하여 반환
Date d = Date.from(instance);
```

- of: 매개변수를 여러개 받아 적합한 타입의 인스턴스를 반환하는 집계 메서드
```java
// 파라미터로 전달 받은 타입의 Enum을 담은 Set을 반환
Set<English> englishTrumpCards = Enumset.of(J, Q, K);
```

- valueOf: from과 of의 더 자세한 버전
```java
BigInteger prime = BigInteger.valueOf(Integer.MAX_VALUE);
```

- instance 혹은 getInstance: 인스턴스를 반환, 그러나 같은 인스턴스임을 보장하지는 않음.
- create 혹은 newInstance: 호출시 매번 새로운 인스턴스 생성하여 반환
- getXXX: getInstance와 같으나, 생성할 클래스가 아닌 다른 클래스에 팩터리 메서드를 정의할 때 사용.
```java
FileStore fileStore = Files.getFileStore(path);
```
- newXXX: newInstance와 같으나, 생성할 클래스가 아닌 다른 클래스에 팩터리메서드를 정의할 때 쓴다.
- XXX: getXXX 와 newXXX의 간결한 버전
```java
List<Complaint> litany = Collections.list(legacyListany);
```

# 핵심 정리 📚
1. 정적 팩터리 메서드 vs public 생성자는 각자 쓰임새가 있다.
2. 상대적인 장단점을 이해하고 `알잘딱깔센` 하게 이용하자.

**Comment by chikeem90** [comment]
추가적으로 동반 클래스 얘기 나올 때 interface에 대한 내용도 나오는데요.
대부분 아시겠지만,, java 8버전부터는 interface에서 static 멤버(+ default 메서드)를 가질 수 있습니다.
https://soft-dino.tistory.com/25 
이거 참고하시면 좋을 거 같아요.
책에도 나오지만 static 멤버가 가능해지면서 Collections 같은 동반클래스를 사용하지 않고도 구현체를 반환하는 정적 팩토리 메서드가 interface에 바로 선언되어질 수 있네요.

**Comment by chikeem90** [comment]
> 2. 정적 팩터리 메서드는 프로그래머가 찾기 어렵다.

이 부분이 사실 좀 이해가 안되었었는데, 백기선님 강의보니 무슨 말인지 이해가 되더라구요.
https://docs.spring.io/spring-boot/docs/current/api/org/springframework/boot/SpringApplication.html
이거처럼 java doc에서 생성자는 따로 구분해서 보여주고 있으나, 정적 팩터리 메서드는 특별히 다뤄주지 않아서? 찾기 어렵다? 정도인 거 같네요

**Reply by YuDeokRin** [reply]
이거 참고 했는데 어렵더군요..

**Comment by chikeem90** [comment, marked as answer]
> 5. 정적 팩터리 메서드를 작성하는 시점에는 반환할 객체의 클래스가 존재하지 않아도 된다.

5번에서는 갑자기 낯선 용어들의 나열이라,, 많이 어렵게 느껴졌는데요
용어는 차치하고 java에서 db랑 연결할 때의 방식을 이해하면 될 거 같습니다
https://sysgongbu.tistory.com/m/95
이 블로그에 잘 나와있는데 제가 이해한바로는 Class.forName에서 string으로 com.mysql.jdbc.Driver (FQCN) 를 받아서 동적으로 클래스 로딩을 하고 있기 때문에, 코드를 작성하는 시점에는 없더라도 추후에 해당 클래스를 만들어서 추가해두면 된다 라고 이해했네요

정리하자면 정적 팩터리 메서드를 작성할 때는 단순 string으로,, (FQCN) 으로 리플렉션 API를 통해서 해당 클래스를 로딩해서 사용할 것이다만 명시해두고 (실제로 해당 클래스가 존재하든 아니든) 이후에 실제 구현부인 클래스를 작성하는 형태로도 가능하다.
https://www.youtube.com/watch?v=X7RXP6EI-5E&list=PLfI752FpVCS8e5ACdi5dpwLdlVkn0QgJJ

이 영상 참고해도 좋습니다.

**Reply by Irisation23** [reply]
@chikeem90 간단하게 정리 해주셔서 감사합니다. 해당 부분이 사실 제가 정리하게 되었던 `아이템 5의 의존 객체 주입`과 유사한 느낌이 들어서 
정리하기 애매 했는데 감사합니다 😀

**Reply by jinan159** [reply]
책 읽으면서 장점 5번이 이해가 잘 안됐었는데, 정리 감사합니다👍

**Comment by chikeem90** [comment]
> 정적 팩터리 메서드의 명명 방식

 실무 하면서도 이런 정적 팩터리 메서드를 작성하는 경우가 많은데요,, 
 항상 네이밍이 고민이 되는 부분인데 정리되어있으니 좋네요,, ㅎㅎ

**Reply by YuDeokRin** [reply]
네이밍 정말 중요한거 같아요.. TDD에서도 마찬가지로 네이밍을 잘해야된다고 하더라고요. 

아래 링크는 다들 아시겠지만 구글 자바 컨벤션 링크입니다. 우O한 코스도 이걸 참고 한다고 들었습니다.
https://google.github.io/styleguide/javaguide.html

**Comment by jinan159** [comment]
1번 아이템 내용중에서 저는 개인적으로 장점 첫 번째(이름을 가질 수 있다)가 가장 와닿았던 것 같아요.
개발하면서 정적 팩터리 메서드로 이름을 부여한것과 그냥 클래스는 코드 읽는 시간을 획기적으로 단축시켜주는게 아주 큰 장점이라 생각합니다.

다른 내용들은 아직 제가 경험이 부족해 겪어본적이 없어서 '아 그렇구나' 정도로만 생각했었는데요.
예제 코드와 함께 설명해주셔서 이해하는데 많은 도움이 되었습니다😄
정리 감사해요👍
